trch/tests/mmu.c

```diff
@@ -18,13 +20,282 @@ int test_rt_mmu()
+    if (mmu_map(trch_ctx, virt_write_addr, phy_read_addr,
+                          mapping_sz)) {
+        printf("test mmu 32-bit access physical map-write-read: mapping create failed\n");
+        goto cleanup_map;
+    }
+
+    mmu_enable(mmu_32);
+
+    old_val = *((uint32_t*)virt_write_addr);
+    *((uint32_t*)virt_write_addr) = val;
+
+    mmu_disable(mmu_32);
+
+    if (*((uint32_t*)phy_read_addr) == val) {
+        success = 1;
+    }
+    else {
+        printf("test mmu 32-bit access physical map-write-read: read wrong value\n");
+        success = 0;
+    }
+
+    mmu_enable(mmu_32);
+
+    *((uint32_t*)virt_write_addr) = old_val;
+
+    mmu_disable(mmu_32);
+
+    mmu_unmap(trch_ctx, virt_write_addr, mapping_sz);
+cleanup_map:
+    mmu_stream_destroy(trch_stream);
+cleanup_stream:
+    mmu_context_destroy(trch_ctx);
+cleanup_context:
+    balloc_destroy(ba);
+cleanup_balloc:
+    mmu_destroy(mmu_32);
+
+    if (success) {
+        return 0;
+    }
+    return 1;
+}
+
+int test_32_mmu_access_physical_wmr(uint32_t virt_read_addr, uint32_t phy_write_addr, unsigned mapping_sz){
+    //write-map-read test
+    //1. TRCH creates a mapping
+    //2. TRCH writes to a low physical address
+    //3. TRCH enables MMU
+    //4. TRCH reads the virtual address corresponding to the physical address written to and checks if data is correct
+
+    //Test is successful if TRCH reads the correct data
+
+
+    _Bool success = 0;
+
+    struct mmu *mmu_32;
+    struct mmu_context *trch_ctx;
+    struct mmu_stream *trch_stream;
+    struct balloc *ba;
+
+    uint32_t val = 0xbeeff00d;
+    uint32_t old_val;
+
+    mmu_32 = mmu_create("RTPS/TRCH->HPPS", RTPS_TRCH_TO_HPPS_SMMU_BASE);
+
+    if (!mmu_32)
+        return 1;
+
+    mmu_disable(mmu_32); // might be already enabled if the core reboots
+
+    ba = balloc_create("32", (uint64_t *)RTPS_HPPS_PT_ADDR, RTPS_HPPS_PT_SIZE);
+
+    if (!ba)
+        goto cleanup_balloc;
+
+    trch_ctx = mmu_context_create(mmu_32, ba, MMU_PAGESIZE_4KB);
+    if (!trch_ctx) {
+        printf("test mmu 32-bit access physical write-map-read: trch context create failed\n");
+        goto cleanup_context;
+    }
+
+    trch_stream = mmu_stream_create(MASTER_ID_TRCH_CPU, trch_ctx);
+    if (!trch_stream) {
+        printf("test mmu 32-bit access physical write-map-read: trch stream create failed\n");
+        goto cleanup_stream;
+    }
+
+    old_val = *((uint32_t*)phy_write_addr);
+    *((uint32_t*)phy_write_addr) = val;
+
+    if (mmu_map(trch_ctx, virt_read_addr, phy_write_addr,
+                          mapping_sz)) {
+        printf("test mmu 32-bit access physical write-map-read: mapping create failed\n");
+        goto cleanup_map;
+    }
+
+
+    mmu_enable(mmu_32);
+
+    if (*((uint32_t*)virt_read_addr) == val) {
+        success = 1;
+    }
+    else {
+        printf("test mmu 32-bit access physical write-map-read: read wrong value\n");
+        success = 0;
+    }
+
+    *((uint32_t*)virt_read_addr) = old_val;
```

Comment:
I think the safest way to restore the memory contents would be to do the reverse of how you set old_val in the first place.  Specifically, it's probably best to first disable the MMU, then set the contents at address phy_write_addr to be equal to old_val.